Only apply WMMA unrolling to elementwise ops of sufficient rank

qedawkins opened the pull request:
In the future WMMA unrolling needs to somehow restrict unrolling to ops that are intended to be later lowered to the opaque WMMA types. For now check the rank of elementwise to prevent crashes when there are flat elementwise vector operations present.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp:
```diff
@@ -556,40 +556,122 @@ Value emitGPUGroupReduction(Location loc, OpBuilder &builder, Value input,
+  for (auto result : op->getResults()) {
+    for (auto user : result.getUsers()) walkSSAChain(user, type, mapping);
+  }
+}
```
Can we use slice utility available upstream?

qedawkins replied:
Not too familiar with the slicing utilities, but from what I can tell it will be tricky to get to work for this case. The tricky part comes when unrolling on a leading fill. Essentially the fill can store to some subview and the the contract will read from a different one, the only connection being that they are both storing/reading the same memref. Thus to find that the fill is "connected" to a contract we need to slice upwards to an earliest shared memref (we should have at least the allocation/binding) and then slice downwards to the fill store.

(https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/tensorcore_vectorization.mlir illustrates this)

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:
```diff
@@ -378,20 +378,21 @@ static void addUnrollVectorsGpuMmaSyncPatterns(RewritePatternSet &patterns) {
-static void addUnrollVectorsGpuWmmaPatterns(RewritePatternSet &patterns) {
+static void addUnrollVectorsGpuWmmaPatterns(
+    RewritePatternSet &patterns, SmallVector<int64_t, 3> &mnkShape,
+    llvm::DenseMap<Operation *, WMMAFragType> &typeMap) {
```
now that I think about it, one of the problem with this is that `Operation*` may be delete and a new operation may be created with the same pointer value. Those are usually very tricky to debug so either we add a listener to make sure we remove delete Operation or we need to  not hold such data structure.  
I also wonder if we can just apply the pattern to the ops we care about?

qedawkins replied:
I see, applying directly to the op list makes sense but it would probably be tricky to get the ordering right (e.g. transfer_reads/ops on the inputs get the shape from the slice coming from the unrolled contract).

ThomasRaoux reviewed (commented):
I think we need a slightly different solution to avoid running into problem of stale pointers. Happy to talk offline if you need help

qedawkins committed:
Only apply WMMA unrolling to elementwise vectors of sufficient rank

In the future WMMA unrolling needs to somehow restrict unrolling to ops
that are intended to be later lowered to the opaque WMMA types. For now
check the rank of elementwise to prevent crashes when there are flat
elementwise vector operations present.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp:
```diff
@@ -587,10 +587,29 @@ std::optional<SmallVector<int64_t>> getWmmaNativeVectorSize(Operation *op) {
-      // Map elementwise ops to the output shape.
-      nativeSize.append({m, n});
-      return nativeSize;
+      if (vecType.getRank() >= 2) {
```
nit: prefer early return style:
```
if (vecType.getRank() >= 2)
  return std::nullopt;
```

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/transform_vector_to_mma.mlir:
```diff
@@ -59,3 +59,166 @@ transform.sequence failures(propagate) {
+hal.executable private @convolution  {
+builtin.module {
+// CHECK-LABEL: func.func @convolution
+func.func @convolution() {
```
please reduce the test and try to only test one thing.

qedawkins replied:
I isolated the rank 1 elementwise + broadcast from the example which was causing crashes. I kept it with checks for subgroup mma ops because I want to make sure that unroll + conversion to mma still works with this present. If wanted I can merge this test with the previous one though because much of it is duplicated.

qedawkins committed:
Simplify test and early exit

qedawkins committed:
clang format

ThomasRaoux reviewed (approved):
Thanks!
